### ISSUE TITLE
useFetch does not work when called inside setup or inside lifecycle hook

### ISSUE THREAD

**Issue by siddharth223437** [nuxt/framework#4548]
### Environment

Operating System: Darwin
Node Version: v17.2.0
Nuxt Version: 3.0.0-27491748.3186841
Package Manager: yarn@1.22.10
Builder: vite
User Config: -
Runtime Modules: -
Build Modules: -

### Reproduction

https://codesandbox.io/s/unruffled-rain-zzuixq?file=/app.vue

### Describe the bug

When using useFetch composable inside setup method or inside any lifecycle hook then it directly shows error. It does not even hit the backend api.

### Logs

```shell
RefImpl {
  __v_isShallow: false,
  dep: undefined,
  __v_isRef: true,
  _rawValue: null,
  _value: null
}
RefImpl {
  __v_isShallow: false,
  dep: undefined,
  __v_isRef: true,
  _rawValue: null,
  _value: null
}
```

**Comment by bbrn** [comment]
I've noticed this issue too. Here are a few other pieces of information that I've noticed that may be helpful in determining the problem:
- this issue happens when the page is refreshed
- the issue does not happen with useFetch if the call is made from a user action, like a button click triggering the api call
- if I use $fetch instead of useFetch, the call is successful, even on page refresh
- the useFetch call returns immediately, even with `await` preceding the call

**Comment by Aareksio** [comment]
The issue is with `server: false` not attempting to invoke `initialFetch`. I am positive this code here is at fault: [asyncData.ts#L75](https://github.com/nuxt/framework/blob/main/packages/nuxt/src/app/composables/asyncData.ts#L75).

```diff
  // Setup hook callbacks once per instance
  const instance = getCurrentInstance()
- if (instance && !instance._nuxtOnBeforeMountCbs) {
+ if (instance && instance._nuxtOnBeforeMountCbs) {
    const cbs = instance._nuxtOnBeforeMountCbs = []
    if (instance && process.client) {
      onBeforeMount(() => {
        cbs.forEach((cb) => { cb() })
        cbs.splice(0, cbs.length)
      })
      onUnmounted(() => cbs.splice(0, cbs.length))
    }
  }
```

Currently the `if` is not entered if `instance._nuxtOnBeforeMountCbs` is defined, but the contents rely on `instance._nuxtOnBeforeMountCbs` being an array.

**Comment by Aareksio** [comment]
Seems to me like [CodeSandbox](https://codesandbox.io/s/angry-wing-tpnmlt?file=/app.vue) is not executing the client-side JavaScript at all. May have something to do with the error below:

![image](https://user-images.githubusercontent.com/11166786/164893007-a14782bd-c2b9-4fd4-ae82-47e097c0aa5c.png)

**Comment by siddharth223437** [comment]
There is an error on server side, see the terminal output. For the first time useFetch is trying to execute api call inside server.

<img width="724" alt="Screen Shot 2022-04-23 at 10 57 54 AM" src="https://user-images.githubusercontent.com/17189936/164913711-1dd6e3e1-5af9-463e-9aa0-9bcd3dcd2641.png">

**Comment by Aareksio** [comment]
This is not an error, this is how your `data` looks when passed to `console.log`.

**Comment by siddharth223437** [comment]
ok, but useFetch is not getting executed on client side. i am getting null output. Though i can add a timeout as a workaround but its not ideal on every page.

Response is not empty. Just grab that api and run on your browser, you will then see the response. Response should look like :

`{ "message": "https://images.dog.ceo/breeds/retriever-flatcoated/n02099267_4546.jpg", "status": "success" }`

If you add a button and execute useFetch then you will the response. 

Attached screenshot is from chrome developer tool.

<img width="751" alt="Screen Shot 2022-04-23 at 12 15 20 PM" src="https://user-images.githubusercontent.com/17189936/164916664-30260f1e-5fa6-45f2-a8d3-ab8d198ae094.png">

**Comment by matthewpull** [comment]
I'm getting this as well since updating to the release candidate (although my previous nuxt3 version was a few months old) - for me, `useLazyFetch` data is undefined on the client side unless I do something like `onMounted(() => refresh())`.

Looking in the console at `window.__NUXT__.data`, the correct payload is there but not being accessed for some reason. Like, the SSR'd HTML coming from the server is correct but once hydration happens the data becomes undefined, which then causes a `Hydration node mismatch` warning.

**Comment by Aareksio** [comment]
@siddharth223437 

You are using `server: false`, which means you must add fallback for server yourself:

```html
<template>
  <div>
    <h1>Random Dog</h1>

    <div v-if="!loading">
      <!-- This is NOT rendered on server, only renders on client after API call completes -->
      <img src="dogImageSrouce"/>
    </div>
  </div>
</template>

<script setup>
const { data, loading } = await useFetch('https://dog-pictures.com/random', { server: false })
const dogImageSrouce = computed(() => data.value ? data.value.message : null)
</script>
```

**Comment by siddharth223437** [comment]
@Aareksio you can try removing server:false. You will get the same error log.

**Comment by Aareksio** [comment]
> @Aareksio you can try removing server:false. You will get the same error log.

@siddharth223437 https://stackblitz.com/edit/github-8lgefy?file=app.vue

**Comment by siddharth223437** [comment]
If I want to perform some logic on returned response then how can I achieve that if nuxt is not blocking call on client side ?

**Comment by Aareksio** [comment]
Please check the example above, it shows a very basic way to protect both `template` and `script` logic. 

It is not blocking on `server`, you **explicitly** pass `server: false`. If you want it to block, the option.

**Comment by siddharth223437** [comment]
@Aareksio You are not getting my point, i don't want to just display response. Please see my below code

```
<template>
  <div>don't display response from api</div>
</template>

<script setup>
const url = 'https://dog.ceo/api/breeds/image/random';
//const options = { server: false };

const { data } = await useFetch(url);
//i want to do something like
// if(data.value){
// perform some logic, at this point i am not getting any result back on client side
//}
</script>
```

**Comment by bbrn** [comment]
I agree with @siddharth223437. The main issue for me is that the call is not blocked and the response never gets updated with the payload. So, the `await` is pointless. For me, the call is made to the backend but just doesn't get blocked and then the only way to get the data is to watch `data`. This ends up behaving like useLazyFetch, which isn't what is expected.

**Comment by danielroe** [comment]
On initial load, with `server: false`, we _can't_ await the data fetching before hydration is complete or we'll actually cause a hydration error (because the client-side rendered state will be different by this point compared to the SSR HTML).

Note that on client-side navigation we do await the fetching as you expect.

Here's the code:

https://github.com/nuxt/framework/blob/4712e994112a97feba447b3edb4afb65e968ed63/packages/nuxt/src/app/composables/asyncData.ts#L136-L154

I agree that this is somewhat counter-intuitive. If you have any suggestions for improving the initial load experience, please do let me know. But for now you will need to handle a pending state if you have set `server: false`.

**Comment by mario-neuhold** [comment]
Same problem when nuxt is set to `ssr: false`: await is not waiting for `useFetch()` to actually fetch the data.

**Comment by siddharth223437** [comment]
@mario-neuhold  I ended up creating composable which is a wrapper on $fetch

import { useRuntimeConfig } from 'nuxt/app'
import {$fetch} from "ohmyfetch";

**composable/useMyFetch.ts**
```
export const useMyFetch = async (req: string, opts?) => {
  const config = useRuntimeConfig()

  return $fetch(req,{
    ...opts,
    baseURL: config.aisAuth.client.baseUrl,
    headers: {
      Accept: 'application/json',
    },
    credentials: 'include',
    async onResponse({ request, response, options }) {
      if(response.status === 401){
          // do something on 401 response
      }

    }
  }).catch((err) => ({error: err}))
}

```

Then you can use **useMyFetch** inside your .vue file

```
async function myFunc(){
    const resp = await useMyFetch('url')
}
```

**Comment by alfa-jpn** [comment]
I used the immediate option.

```typescript
const fetch = useFetch(url, { immediate: false })
await fetch.execute({ _initial: true })
```

**Comment by cyruscollier** [comment]
Thanks, @alfa-jpn that technique solved this for me (for now!). 

I'll add that in our case, we aren't setting `server: false` explicitly, but instead are calling `useFetch()` within `onMounted()`, which perhaps results in the same behavior. Our application was running on rc.6 for a long time, where calling fetch in `onMounted()` was working fine, and we are just now upgrading to 3.0 stable when this issue surfaced.

@danielroe is there a better way to wait until hydration is finished before making the fetch? In our application we are loading a lot of data on some pages so are splitting the fetching between server and client side to get a better initial response time while still getting the benefits from the most critical data fetched server side.

**Comment by danielroe** [comment]
Ah. If you are calling something within a method, you should almost certainly use `$fetch` directly rather than `useFetch`, which is only meant to be called in the setup method.

**Comment by cyruscollier** [comment]
The documentation says that `useFetch()` can also be used in lifecycle hooks. I think my confusion (and perhaps others) is around the best practice that composable functions should only be called directly at the top level of the setup call stack, not within helper functions or other shared code not directly scoped to setup. Is that right?

**Comment by danielroe** [comment]
@cyruscollier Yes, that's right.
